File: cpp/Lec 2 Build-up Logical Thinking (Patterns)/pattern_12.cpp
// 1      1
// 12    21
// 123  321
// 12344321

//----------------------------------
// BRUTE FORCE APPROACH (Nested loops)
//----------------------------------
#include <iostream>
#include <string>
using namespace std;

int main()
{
    int n = 4;
    for (int i = 1; i <= n; i++)
    {
        // Left part: 1 to i
        for (int j = 1; j <= i; j++)
        {
            cout << j;
        }
        // Middle spaces: 2*(n-i)
        for (int j = 1; j <= 2 * (n - i); j++)
        {
            cout << " ";
        }
        // Right part: i to 1
        for (int j = i; j >= 1; j--)
        {
            cout << j;
        }
        cout << endl;
    }
    return 0;
}

//----------------------------------
// BETTER APPROACH (String building per row)
//----------------------------------
#include <iostream>
#include <string>
using namespace std;

int main()
{
    int n = 4;
    for (int i = 1; i <= n; i++)
    {
        string left, right;
        // Build left and right parts
        for (int j = 1; j <= i; j++)
        {
            left += to_string(j);
            right = to_string(j) + right;
        }
        string spaces(2 * (n - i), ' ');
        cout << left << spaces << right << endl;
    }
    return 0;
}

//----------------------------------
// BEST APPROACH (Pre-allocated buffer with direct indexing)
//----------------------------------
#include <iostream>
#include <string>
using namespace std;

int main()
{
    int n = 4;
    const int width = 2 * n;
    string line(width, ' ');

    for (int i = 1; i <= n; i++)
    {
        // Fill left part
        for (int j = 1; j <= i; j++)
        {
            line[j - 1] = '0' + j;
        }
        // Fill right part (mirror)
        for (int j = 1; j <= i; j++)
        {
            line[width - j] = '0' + j;
        }
        cout << line << '\n';
        // Reset line to spaces
        fill(line.begin(), line.end(), ' ');
    }
    return 0;
}

File: python/Lec 2 Build-up Logical Thinking (Patterns)/pattern_12.py
# 1      1
# 12    21
# 123  321
# 12344321

#----------------------------------
# BRUTE FORCE APPROACH
#----------------------------------
n = 4
for i in range(1, n + 1):
    # Left part
    line = ''.join(str(j) for j in range(1, i + 1))
    # Middle spaces
    line += ' ' * (2 * (n - i))
    # Right part (reverse)
    line += ''.join(str(j) for j in range(i, 0, -1))
    print(line)

#----------------------------------
# BETTER APPROACH (String multiplication)
#----------------------------------
n = 4
for i in range(1, n + 1):
    left = ''.join(str(j) for j in range(1, i + 1))
    right = ''.join(str(j) for j in range(i, 0, -1))
    print(f"{left}{' ' * (2 * (n - i))}{right}")

#----------------------------------
# BEST APPROACH (Single generator expression)
#----------------------------------
n = 4
width = 2 * n
for i in range(1, n + 1):
    # Create row with list comprehension
    row = [' '] * width
    # Fill left and right
    for j in range(1, i + 1):
        row[j - 1] = str(j)
        row[width - j] = str(j)
    print(''.join(row))